### PR TITLE
Realtime verification reporting

### PR DESCRIPTION
**Pull request by MikaelMayer**
This PR enables Boogie to report in real-time its current execution state.
Specifically, it enables the output printers to
* `ReportImplementationsBeforeVerification`: Be notified about the implementations to be verified, for example to change priorities on the fly
* `ReportStartVerifyImplementation`: Be notified when an implementation starts to verify
* `ReportEndVerifyImplementation`: Be notified when an implementation finishes to verify
* `ReportSplitResult`: Be notified when an split finishes
  * There are three places in which I needed to call this: 1) When the prover did not fail, 2) When the prover returns a time out and it was the last change to split the assertions 3) When there is no way it can split the assertions further
  * I changed `VCResult` a bit so that stores the counter-examples obtained from the split itself.
  * I added a method to `VCResult` so that it can compute per-assert outcomes and counter-examples, if any.
* I fixed all the tests on Windows as well (a log file was not closed before being accessed, as a null pointer exception that was throwing because Process was null)
* I added a test for the new behavior of the `OutputPrinter`

** Breaking change** The output printer's interface is changed.

### CODE REVIEW

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (MikaelMayer) [inline Source/ExecutionEngine/ExecutionEngine.cs:812]:
```suggestion
      // Ask the printer for custom priorities to verify implementations
      var userDefinedPriorities =
```

File: Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs
Comment (MikaelMayer) [inline Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs:648]:
```suggestion
```

File: Source/ExecutionEngine/ConsolePrinter.cs
Comment (atomb) [inline Source/ExecutionEngine/ConsolePrinter.cs:198]: I'm wondering whether `perAssertOutcome` should instead be a `VCResult`, as those are what make up the `VerificationResult` passed into `ReportEndVerifyImplementation`.

Also note that there's an `OnVCResult` method in `VerifierCallback`.

File: Source/VCGeneration/Counterexample.cs
Comment (atomb) [inline Source/VCGeneration/Counterexample.cs:520]: Nice simplification.

File: Source/VCGeneration/Split.cs
Comment (atomb) [inline Source/VCGeneration/Split.cs:90]: I think there's already an `Asserts` property containing exactly this.

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (atomb) [inline Source/VCGeneration/SplitAndVerifyWorker.cs:156]: All the code in here will be simpler, I think, if you use the signature I suggested above.

The logic you have for identifying which assertions in a batch failed, based on the counterexamples, will still need to happen somewhere, though.

You may need to merge or rebase to incorporate the current `master` before the `VCResult` will be conveniently available.

**Review by atomb — CHANGES_REQUESTED** [review]
This looks great! I'm wondering whether one of the arguments of `ReportAssertionBatchResult` should be different to take into account recent changes, though. I'm not 100% sure that's the right way to go, but I think it's worth at least considering.

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/SplitAndVerifyWorker.cs:163]: Why does each Split need to look at counterExamples collected by other splits?

I would think that each split reports a CounterExample if it failed, and the CounterExample can say which statements where involved in the failure, which we can then mark red in the IDE if that CounterExample is active.

Btw, I saw that in the IDE you also mark asserts green. Do you mark them green if the assertion passed for each split in which it occurred ? What do you do if an assertion passes for the first completed split, but fails for the split that completes after it?

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/SplitAndVerifyWorker.cs:156]: Relying on `callback` being a specific subtype seems unreliable.

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (keyboardDrummer) [inline Source/ExecutionEngine/ExecutionEngine.cs:29]: I'd leave out the comment.

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (keyboardDrummer) [inline Source/ExecutionEngine/ExecutionEngine.cs:34]: This method doesn't match the name of the interface `OutputPrinter`. Why do you need this to begin with? Can't you use the `{:priority}` attribute on Implementations to control their priority?

File: Source/ExecutionEngine/ConsolePrinter.cs
Comment (MikaelMayer) [reply Source/ExecutionEngine/ConsolePrinter.cs:198]: Which of these do you have in mind?
- `Dictionary<AssertCmd, VCResult> perAssertOutcome`
- `VCResult perAssertOutcome`

`The first one would contain more information so it's definitely not a big deal, I can make it happily.
The second one would require to store more information in VCResult, which I can do, but I am wondering first if this is what you meant.`

File: Source/VCGeneration/Split.cs
Comment (MikaelMayer) [reply Source/VCGeneration/Split.cs:90]: Oh I see, it's in the newly VCsResult.. Let me check how I can use it there.

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (MikaelMayer) [reply Source/VCGeneration/SplitAndVerifyWorker.cs:163]: This code is invoked after all the counter-examples of that split have been emitted. However, counter-examples are not recorded per split result (not even in the current master), so I have to rely on the entire list of previously emitted counter-examples, and disregarding counter-examples that were emitted by other splits. Let me know how I can do better, e.g. maybe store the counter-examples in VCResult?
Each split reports up to 5 counter examples.

To my understanding, I think that splits are all disjoints, so if an assert happens in one split, it shouldn't appear in another split. Yet I don't rely on that anyway. If a single assert would appear in two different splits, they will be simply be recorded as two different assertions at the same place. Hence, if one of them fail, the error will be given priority on the gutter - the gutter rendering is like a z-buffer where errors decorations are "closer" than verified decorations.

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (MikaelMayer) [reply Source/VCGeneration/SplitAndVerifyWorker.cs:156]: Agreed. I think we should put the counter-examples in the VCResult so that I don't need to rely on the collector itself.

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (MikaelMayer) [reply Source/ExecutionEngine/ExecutionEngine.cs:34]: What do you mean by "doesn't match the name of the interface `OutputPrinter`"?

I need this method so that if a particular output printer is temporarily interested by the result of verifying a specific implementation first, it can set a priority independent of the source code, e.g. for an IDE extension where the user wants to verify the last edited implementation.
You can still use the `{:priority}` attribute on implementations to manually control their priority.

One debatable question is, how to merge a manual and an output-printer given priority. In this PR, I give priority to the output printer - in the IDE extension I'm developing, it will always verify the last five implementations. We could also max the priorities. But I found that, in an IDE setting, it makes it possible to choose the order of verification without writing explicit `{:priority}` attributes, which is pretty neat.

File: Source/ExecutionEngine/ConsolePrinter.cs
Comment (atomb) [reply Source/ExecutionEngine/ConsolePrinter.cs:198]: I'm thinking the latter, which includes all of the assertions in the batch. It only includes one outcome, though, so you'll still need the logic you have (somewhere) to use the counterexamples to figure out which of the assertions in a batch failed when the outcome indicates failure.

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (atomb) [reply Source/VCGeneration/SplitAndVerifyWorker.cs:163]: I think it may make sense, as you're beginning to suggest here, to adapt `VCResult` to include counterexamples and multiple outcomes (using the logic you implement in this PR). That will require a lot less work from clients of that type in most cases, I expect.

File: Source/ExecutionEngine/ConsolePrinter.cs
Comment (MikaelMayer) [reply Source/ExecutionEngine/ConsolePrinter.cs:198]: I'll try to store the counter-examples in this VCResult so that it makes it easier.

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (keyboardDrummer) [reply Source/ExecutionEngine/ExecutionEngine.cs:34]: I understand why you added the method, but why don't you use the priority attribute instead?

File: Source/ExecutionEngine/ConsolePrinter.cs
Comment (MikaelMayer) [reply Source/ExecutionEngine/ConsolePrinter.cs]: I ended up storing the counter-examples in the VCResult, and the code to figure out per-assertion outcome is now in a method of this VCResult that can be invoked.

File: Source/VCGeneration/Split.cs
Comment (MikaelMayer) [reply Source/VCGeneration/Split.cs]: Removed in the last commit

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (MikaelMayer) [reply Source/ExecutionEngine/ExecutionEngine.cs]: Aahhhh ok got it got it. I historically discovered the {:priority} attribute while writing this code, but I didn't know at that time I could change the implementation, I was trying to only modify the verification behavior. But now you suggestion makes sense to me. I implemented in my branch something where it sets priorities of implementations in the newly added callback, so that there is no need to order them later.
Everything is working ! thanks for the suggestion.

**Comment by MikaelMayer** [comment]
I think I have a much cleaner solution now.

**Review by atomb — DISMISSED** [review]
Looks great!

**Comment by MikaelMayer** [comment]
> Looks great!

I will wait for at least the approval of @keyboardDrummer to merge this PR.

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (keyboardDrummer) [inline Source/ExecutionEngine/ExecutionEngine.cs:182]: It's moved into ExecutionEngineOptions in this PR: https://github.com/boogie-org/boogie/pull/520

Could you remove the comment?

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (keyboardDrummer) [inline Source/ExecutionEngine/ExecutionEngine.cs:788]: This comment seems obsolete

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (keyboardDrummer) [inline Source/ExecutionEngine/ExecutionEngine.cs:971]: Do you also want to report this if a cached result was found? If not, please move this call inside the `if`.

Note that `ReportEndVerifyImplementation` is only called when no cached result is found.

I recommend having the `ReportEndVerifyImplementation` and `ReportStartVerifyImplementation` calls in the same method, so it's easier to see that both are called in the same code path.

File: Source/VCGeneration/Counterexample.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/Counterexample.cs:511]: Would `FailingAssert` be a more specific name? I'm not able to get any information out of the term `Underlying`

File: Source/VCGeneration/Counterexample.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/Counterexample.cs:691]: Why don't you do place `split.Counterexamples.Add(ce)` before the call to `OnCounterexample` in `SplitAndVerifyWorker`, then you don't need this class.

File: Source/VCGeneration/IConditionGenerationLogger.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/IConditionGenerationLogger.cs:7]: Newline seems unnecessary

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/SplitAndVerifyWorker.cs:165]: `callback` variable is shadowing the field variable.

File: Source/VCGeneration/IConditionGenerationLogger.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/IConditionGenerationLogger.cs:7]: Consider renaming this to `LogSplitResult` since it's a method in the a `Logger` interface.

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/SplitAndVerifyWorker.cs:105]: Could you move the logic for recording which CounterExamples a Split has into the `Split` class? `Split` creates a `VCGen.ErrorReporter` where this callback is also passed.

File: Source/ExecutionEngine/ConsolePrinter.cs
Comment (MikaelMayer) [reply Source/ExecutionEngine/ConsolePrinter.cs:194]: Oops, sorry, looks like a leftover !

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (MikaelMayer) [reply Source/ExecutionEngine/ExecutionEngine.cs:971]: Yes we want to report this even if a cached result was found. When we invoke Boogie to verify a procedure, it's an internal detail whether it uses the cache or prove it again.
My extension needs the feedback that verification was started and finished for a given implementation, even if there is no split inside.

Good catch, I moved `ReportEndVerifyImplementation ` in the same method as `ReportStartVerifyImplementation `. You actually just discovered a potential bug which did happen to me many times, when my extension reports that something never finishes to verify but Boogie is done.

File: Source/VCGeneration/Counterexample.cs
Comment (MikaelMayer) [reply Source/VCGeneration/Counterexample.cs:511]: Makes sense

File: Source/VCGeneration/Counterexample.cs
Comment (MikaelMayer) [reply Source/VCGeneration/Counterexample.cs:691]: You mean, before this call?
`callback.OnCounterexample(split.ToCounterexample(split.Checker.TheoremProver.Context), msg);`
The problem is, `OnCounterExample` is not called only from `SplitAndVerifyWorker`. It's also called from `VCGen` 2 times. Look for `ErrorReporter.OnModel`, it uses the callback defined in `Split.cs` line `reporter = new VCGen.ErrorReporter(...`, which callback is a `SplitVerifyCallback` that records the counter-examples in the split as you have seen.

File: Source/VCGeneration/IConditionGenerationLogger.cs
Comment (MikaelMayer) [reply Source/VCGeneration/IConditionGenerationLogger.cs:7]: Renamed instead the interface to `IConditionGenerationPrinter` since it's extended by `OutputPrinter`

File: Source/VCGeneration/SplitAndVerifyWorker.cs
Comment (MikaelMayer) [reply Source/VCGeneration/SplitAndVerifyWorker.cs:105]: I wish I could move this logic, however, it would require me to also move the logic for handling the split's result into the split itself, since `SplitAndVerifyWorker.HandleProverFailure` and `SplitAndVerifyWorker.ProcessResult` both rely on the `SplitVerifierCallback`. Let me know if you would like me to attempt such a refactoring as part of this PR, or if it is fine for now.

Comment (MikaelMayer) [reply Source/VCGeneration/SplitAndVerifyWorker.cs]: Looking at your solution, I was able to find another one that did not even need to change the line you pointed at.
The class SplitVerifierCallback is now removed !

**Review by keyboardDrummer — DISMISSED** [review]

**Comment by MikaelMayer** [comment]
* I fixed all the tests on Windows as well (a log file was not closed before being accessed, as a null pointer exception that was throwing because Process was null)
* I added a test for the new behavior of the `OutputPrinter`

File: Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs
Comment (keyboardDrummer) [inline Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs:32]: When does this happen?

File: Source/VCGeneration/Split.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/Split.cs:1563]: Why would be close here? Shouldn't the theoremProver of the checker be reused?

File: Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs:32]: I have no idea but it did happen on my machine and caused a crash as I was debugging the Dafny Language Server and heavily loaded it with a lot of requests.
The only reasons why `Process` would be null would be
- If it `GoBackToIdle()` was called before a single call to `SetupProcess()`
- If the method `Close()` was called before `GoBackToIdle()`

It would be worth investigating why, but for now, I found it should be sound to do this change, because if the process is null, it means it was either not started or it was closed, so it can always be restarted and the prover can be reused.

File: Source/VCGeneration/Split.cs
Comment (MikaelMayer) [reply Source/VCGeneration/Split.cs:1563]: Good question, let me tell you why this is needed in my case, and so we can discuss alternatives.
When I ran the boogie tests on Windows, it kept crashing 90% of the tests on `GetProverLog.cs` on line 35 because it's trying to read the log files and these logs files were being written by the theorem prover and never closed (`StreamWriter`). So, somewhere I need to close these log files and this was never done explicitely except in the public `Close()` method of the theorem prover which was not call.
Any suggestions to do otherwise?

Comment (keyboardDrummer) [reply Source/VCGeneration/Split.cs:1563]: The trick is to Dispose the ExecutionEngine before reading the log files, like it is done in Dafny [here](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyPipeline.Test/IntraMethodVerificationStability.cs#L217).

Comment (MikaelMayer) [reply Source/VCGeneration/Split.cs:1563]: That's a good trick ! It works !

**Review by keyboardDrummer — APPROVED** [review]

**Comment by MikaelMayer** [comment]
I managed to solve all the merging conflicts; however, I cannot push my changes as a few early adopters are relying on both my Dafny branch and Boogie branch to be in sync. @keyboardDrummer please let me know when Dafny will be compatible with the newest version of Boogie.

File: Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs
Comment (MikaelMayer) [reply Source/Provers/SMTLib/SMTLibInteractiveTheoremProver.cs:32]: Just added a test that fails by replicating the behavior "close()" before "GoBackToIdle()". It's not the panacea, as in the first place we should never have such a call and there might be some more lurking bugs. But it's better than nothing.

File: Source/VCGeneration/Split.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/Split.cs:36]: Don't all the counterexamples in `Split.CounterExamples` belong to this Split after your latest changes? Does the continue on line 49 ever trigger?

File: Source/VCGeneration/Split.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/Split.cs:56]: This part is a bit confusing to me. I get line 58, but line 59 I'm not sure about. Shouldn't that be `ProverInterface.Outcome.Undetermined`?

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (MikaelMayer) [reply Source/ExecutionEngine/ExecutionEngine.cs:988]: On the first sight, `ExecutionEngine.Options` is a `ExecutionEngineOptions`, whereas `CheckerPool.Options` is a `VCGenOptions`, which have nothing in common (and not the printer which is only in `ExecutionEngineOptions`).
Let me know if you still prefer to attach the printer to `CheckerPool.Options`

File: Source/VCGeneration/Split.cs
Comment (MikaelMayer) [reply Source/VCGeneration/Split.cs:36]: Normally they should. If we were to write this in Dafny, I would just put some asserts. this is just a guard now in case something went wrong. I'll work on the comment.

File: Source/VCGeneration/Split.cs
Comment (MikaelMayer) [reply Source/VCGeneration/Split.cs:56]: You have a point, let me explain the other one.
`ProverInterface.Outcome.Undetermined` was my first choice too originally, but then I figured that if we arrive at this point, then either `outcome != ProverInterface.Outcome.Invalid` or `counterExamples.Count >= maxCounterExamples`
Hence.
1. If `outcome != ProverInterface.Outcome.Invalid`, then because `outcome != ProverInterface.Outcome.Valid` as well (we are on the else branch), the remaining outcomes are either `TimeOut`, `OutOfMemory`, `OutOfResource` or `Undetermined` (I'm not sure what `Bounded` means), which are all more precise than just `Undetermined`, and thus it's nice to report such information at the assertion level.
2. Now if outcome == ProverInterface.Outcome.Invalid, then `counterExamples.Count >= maxCounterExamples`, it means we would assign `ProverInterface.Outcome.Invalid` to the remaining assertions, which is arguably incorrect, so let me change this particular case to `Undetermined` instead.

**Comment by MikaelMayer** [comment]
I added also some missing test portion to at least test the method `vcResult.ComputePerAssertOutcomes`

File: Source/ExecutionEngine/ExecutionEngine.cs
Comment (keyboardDrummer) [reply Source/ExecutionEngine/ExecutionEngine.cs]: > (and not the printer which is only in ExecutionEngineOptions)

I see. Is it possible to put a `IConditionGenerationPrinter Printer {get;}` in `VCGenOptions` which then is automatically filled with the right printer?

**Review by keyboardDrummer — DISMISSED** [review]

File: Source/VCGeneration/VCGenOptions.cs
Comment (keyboardDrummer) [inline Source/VCGeneration/VCGenOptions.cs]: Please resolve the missing space warnings

**Review by keyboardDrummer — DISMISSED** [review]

**Review by keyboardDrummer — APPROVED** [review]

**Pull request merged**
